### PR TITLE
BUG: sctypeDict['f8'] randomly points to double or longdouble depending on non-deterministic dictionary iteration order

### PR DESCRIPTION
**Pull request by eric-wieser**

### CODE REVIEW

File: numpy/core/tests/test_numerictypes.py [hunk 1]
```diff
@@ -406,3 +406,9 @@ def test_sibling_class(self):
         for w1, w2 in itertools.product(self.wrappers, repeat=2):
             assert_(not np.issubdtype(w1(np.float32), w2(np.float64)))
             assert_(not np.issubdtype(w1(np.float64), w2(np.float32)))
+
+
+def TestSctypeDict(object):
+    def test_longdouble(self):
+        assert_(np.sctypeDict['f8'] is not np.longdouble)
+        assert_(np.sctypeDict['c16'] is not np.clongdouble)
```
Comment (eric-wieser) [inline numpy/core/tests/test_numerictypes.py]: This latter test fails for me in 2.7

File: numpy/core/numerictypes.py [hunk 2]
```diff
@@ -323,31 +323,37 @@ def _add_aliases():
...
+        sctypeNA[info.type] = na_name
+        sctypeNA[info.char] = na_name
 
-                # add forward, reverse, and string mapping to numarray
-                sctypeNA[na_name] = info.type
-                sctypeNA[info.type] = na_name
-                sctypeNA[info.char] = na_name
-        if char != '':
-            sctypeDict[char] = info.type
-            sctypeNA[char] = na_name
```
Comment (eric-wieser) [inline numpy/core/numerictypes.py]: This was conditional on the above `if` running anyway, to devine `na_name

Comment (mhvk) [reply numpy/core/numerictypes.py]: nice catch

File: numpy/core/tests/test_numerictypes.py (hunk 1 above)
Comment (mhvk) [reply numpy/core/tests/test_numerictypes.py]: In current master, right?

**Review by mhvk — APPROVED** [review]
Looks good, nice cleanup

File: numpy/core/tests/test_numerictypes.py (hunk 1 above)
Comment (eric-wieser) [reply numpy/core/tests/test_numerictypes.py]: Yes

File: numpy/core/numerictypes.py [hunk 3]
```diff
@@ -323,31 +323,37 @@ def _add_aliases():
...
+        sctypeDict[myname] = info.type
+        sctypeDict[na_name] = info.type
 
-                # add mapping for both the bit name and the numarray name
-                sctypeDict[myname] = info.type
-                sctypeDict[na_name] = info.type
+        # add forward, reverse, and string mapping to numarray
+        sctypeNA[na_name] = info.type
+        sctypeNA[info.type] = na_name
+        sctypeNA[info.char] = na_name
```
Comment (eric-wieser) [inline numpy/core/numerictypes.py]: A bug for a later PR: These two lines should be run unconditionally.

**Comment by mattip** [comment]
Is there a reason to keep `sctypeNA` and `typeNA`? If so they should be documented interfaces. I would prefer to remove as many of these confusing mappings as possible

**Comment by mattip** [comment]
related to issue #11341

**Pull request merged**